app/presenters/facet_presenter.rb

```diff
@@ -0,0 +1,181 @@
+##
+# Display logic for facets
+class FacetPresenter
+  include FacetsHelper
+  include UrlHelper
+  include ActionView::Helpers::NumberHelper
+  delegate :t, to: I18n
+
+  def initialize(facet, controller, configuration = controller.blacklight_config)
+    @facet = facet
+    @configuration = configuration
+    @controller = controller
+    @response = controller.instance_variable_get(:@response)
+  end
+
+  def display(show_children = false)
+    if facet_config.hierarchical && !facet_config.parent
+      hierarchical_facet
+    elsif !facet_config.parent || show_children
+      if facet_config.boolean
+        boolean_facet
+      elsif facet_config.colour
+        colour_facet
+      elsif facet_config.range
+        range_facet
+      else
+        simple_facet
+      end
+    end
+  end
+
+  def facet_item_url(item)
+    if facet_in_params?(@facet.name, item)
+      search_action_url(remove_facet_params(@facet.name, item, @controller.params))
+    else
+      search_action_url(add_facet_params_and_redirect(@facet.name, item))
+    end
+  end
+
+  def facet_config
+    @facet_config ||= @configuration.facet_fields[@facet.name]
+  end
+
+  def basic_facet(options = {})
+    options = options.reverse_merge(type: :simple, count: 4)
+    {
+      title: facet_label(@facet.name),
+      select_one: facet_config.single,
+      items: @facet.items[0..(options[:count] - 1)].map { |item| send(:"#{options[:type]}_facet_item", item) },
+      extra_items: @facet.items.size <= options[:count] ? nil : {
+        items: @facet.items[options[:count]..-1].map { |item| send(:"#{options[:type]}_facet_item", item) }
+      }
+    }.merge(options[:type] => true)
+  end
+
+  def basic_facet_item(item)
+    {
+      url: facet_item_url(item),
+      text: facet_label(@facet.name, item.value),
+      num_results: number_with_delimiter(item.hits),
+      is_checked: facet_in_params?(@facet.name, item)
+    }
+  end
+
+  def simple_facet
+    basic_facet(type: :simple)
+  end
+
+  def simple_facet_item(item)
+    basic_facet_item(item)
+  end
+
+  def colour_facet
+    basic_facet(type: :colour, count: 16)
+  end
+
+  def colour_facet_item(item)
+    basic_facet_item(item).tap do |basic|
+      basic.delete(:text)
+      basic[:hex] = item.value
+    end
+  end
+
+  def hierarchical_facet
+    basic_facet(type: :hierarchical)
+  end
+
+  def hierarchical_facet_item(item)
+    child_facets = hierarchical_facet_item_children(item)
+
+    {
+      has_subfilters: child_facets.present?,
+      filters: child_facets.map { |child|
+        self.class.new(child, @controller).display(true)
+      }
+    }.merge(simple_facet_item(item))
+  end
+
+  def hierarchical_facet_item_children(item)
+    facets_from_request(facet_field_names).select do |child|
+      parent = @configuration.facet_fields[child.name].parent
+      !parent.nil? && (parent.first == @facet.name) && (parent.last == item.value)
+    end
+  end
+
+  def range_facet
```

Comment:
Assignment Branch Condition size for range_facet is too high. [37.34/15]<br>Method has too many lines. [35/10]